File: clients/C++/1.1 mcp_logout/MCPLogoutClient.cpp
#include <iostream>
#include <string>
#include <map>
#include <stdexcept>
#include <json/json.h> // Für JSON-Bibliothek
#include <winsock2.h>
#include <ws2tcpip.h>

#pragma comment(lib, "ws2_32.lib") // Winsock-Bibliothek verlinken

// Argumente parsen
std::map<std::string, std::string> parseArguments(int argc, char* argv[]) {
    std::map<std::string, std::string> args;
    for (int i = 1; i < argc; i++) {
        std::string key = argv[i];
        if (i + 1 < argc) {
            args[key] = argv[++i];
        }
    }
    return args;
}

// Funktion zum Senden der Logout-Anfrage
std::string sendLogoutRequest(const std::string& serverIp, int serverPort, const std::string& token) {
    // JSON-Payload erstellen
    Json::Value payload;
    payload["command"] = "logout";
    payload["token"] = token;

    Json::StreamWriterBuilder writer;
    std::string payloadJson = Json::writeString(writer, payload);

    // Winsock initialisieren
    WSADATA wsaData;
    if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0) {
        throw std::runtime_error("Failed to initialize Winsock.");
    }

    // Socket erstellen
    SOCKET sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock == INVALID_SOCKET) {
        WSACleanup();
        throw std::runtime_error("Failed to create socket.");
    }

    // Server-Adresse konfigurieren
    sockaddr_in serverAddr;
    serverAddr.sin_family = AF_INET;
    serverAddr.sin_port = htons(serverPort);
    if (inet_pton(AF_INET, serverIp.c_str(), &serverAddr.sin_addr) <= 0) {
        closesocket(sock);
        WSACleanup();
        throw std::runtime_error("Invalid server IP address.");
    }

    // Verbindung herstellen
    if (connect(sock, (sockaddr*)&serverAddr, sizeof(serverAddr)) < 0) {
        closesocket(sock);
        WSACleanup();
        throw std::runtime_error("Connection failed.");
    }

    // Daten senden
    if (send(sock, payloadJson.c_str(), payloadJson.size(), 0) < 0) {
        closesocket(sock);
        WSACleanup();
        throw std::runtime_error("Failed to send data.");
    }

    // Antwort empfangen
    char buffer[4096];
    int bytesRead = recv(sock, buffer, sizeof(buffer) - 1, 0);
    if (bytesRead < 0) {
        closesocket(sock);
        WSACleanup();
        throw std::runtime_error("Failed to receive data.");
    }

    buffer[bytesRead] = '\0'; // Antwort null-terminieren

    // Socket schließen
    closesocket(sock);
    WSACleanup();

    return std::string(buffer);
}

int main(int argc, char* argv[]) {
    try {
        auto args = parseArguments(argc, argv);

        // Erforderliche Parameter extrahieren
        std::string serverIp = args["--server-ip"];
        int serverPort = std::stoi(args["--server-port"]);
        std::string token = args["--token"];

        if (serverIp.empty() || serverPort == 0 || token.empty()) {
            std::cerr << "❌ ERROR: Missing required parameters.\n";
            return 1;
        }

        std::cout << "🔒 Sending logout request...\n";

        // Logout-Anfrage senden und Antwort erhalten
        std::string response = sendLogoutRequest(serverIp, serverPort, token);

        std::cout << "Response from server:\n" << response << "\n";
    } catch (const std::exception& e) {
        std::cerr << "❌ ERROR: " << e.what() << "\n";
        return 1;
    }

    return 0;
}

File: clients/Java/1.1 mcp_logout/MCPLogoutClient.java
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

public class MCPLogoutClient {

    public static void main(String[] args) {
        String serverIp = null;
        int serverPort = 0;
        String token = null;

        // Argumente parsen
        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--server-ip":
                    serverIp = args[++i];
                    break;
                case "--server-port":
                    serverPort = Integer.parseInt(args[++i]);
                    break;
                case "--token":
                    token = args[++i];
                    break;
            }
        }

        // Überprüfen, ob alle Argumente angegeben wurden
        if (serverIp == null || serverPort == 0 || token == null) {
            System.out.println("❌ ERROR: Missing required parameters.");
            return;
        }

        System.out.println("🔒 Sending logout request...");
        String response = sendLogoutRequest(serverIp, serverPort, token);
        System.out.println("Response from server:");
        System.out.println(response);
    }

    private static String sendLogoutRequest(String serverIp, int serverPort, String token) {
        // JSON-Payload erstellen
        JSONObject payload = new JSONObject();
        payload.put("command", "logout");
        payload.put("token", token);

        // In String umwandeln
        String payloadJson = payload.toString();

        try (Socket client = new Socket(serverIp, serverPort)) {
            // OutputStream holen und Daten senden
            OutputStream out = client.getOutputStream();
            byte[] data = payloadJson.getBytes(StandardCharsets.UTF_8);
            out.write(data);
            out.flush();

            // Antwort empfangen
            InputStream in = client.getInputStream();
            byte[] buffer = new byte[4096];
            int bytesRead = in.read(buffer);
            if (bytesRead == -1) {
                return "❌ ERROR: No response from server.";
            }

            return new String(buffer, 0, bytesRead, StandardCharsets.UTF_8);

        } catch (IOException ex) {
            return "Error: " + ex.getMessage();
        }
    }
}
